core/operations.onyx

```diff
@@ -13,15 +15,30 @@ package builtin
 
 //
 // Allows for basic array programming support
-#operator +  macro (l, r: [$N]$T) => __array_math_op(l, r, [a, b](a + b));
-#operator -  macro (l, r: [$N]$T) => __array_math_op(l, r, [a, b](a - b));
-#operator *  macro (l, r: [$N]$T) => __array_math_op(l, r, [a, b](a * b));
-#operator /  macro (l, r: [$N]$T) => __array_math_op(l, r, [a, b](a / b));
-#operator == macro (l, r: [$N]$T) => core.intrinsics.wasm.memory_equal(cast(rawptr)l, cast(rawptr)r, N * sizeof T);
+
+// Array w/ array operations
+#operator + macro (l, r: [$N]$T) => __array_op_array(l, r, [a, b](a + b));
+#operator - macro (l, r: [$N]$T) => __array_op_array(l, r, [a, b](a - b));
+#operator * macro (l, r: [$N]$T) => __array_op_array(l, r, [a, b](a * b));
+#operator / macro (l, r: [$N]$T) => __array_op_array(l, r, [a, b](a / b));
+
+// Array w/ scalar operations
+#operator + macro (l: [$N]$T, r: T) => __array_op_scalar(l, r, [a, b](a + b));
+#operator - macro (l: [$N]$T, r: T) => __array_op_scalar(l, r, [a, b](a - b));
+#operator * macro (l: [$N]$T, r: T) => __array_op_scalar(l, r, [a, b](a * b));
+#operator / macro (l: [$N]$T, r: T) => __array_op_scalar(l, r, [a, b](a / b));
```

Comment:
Do we also want operations for scalar /w array (ie. `2 * i32.[ 1, 2, 3 ] == [ 2, 4, 6 ]`)? Would probably make this feel more cohesive

Reply:
I have wanted to add a `#commutative` directive that can go on operator overloads that would automatically try both A + B and B + A, so that would solve this rather nicely.

Reply:
Perfect, I can roll that into this PR if you'd like. Assuming this would be a flag on `AstDirectiveOperator` and a bit of logic added to overload resolution?

Reply:
I can handle this one. Its gonna take a bit of work because of how I'm currently storing operator overloads. They are currently no different than overloaded functions after the `AstDirectiveOperator` is processed, so I think I will have to modify a quite a few things to make it work. Unless you've already done it, then that's okay!

Reply:
Sounds good! I've only done some basic work towards that, so I didn't sink too much time into it.